py/torch_tensorrt/dynamo/_engine_caching.py

```diff
@@ -0,0 +1,268 @@
+import copy
+import logging
+import os
+import pickle
+import shutil
+import sys
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List, Optional, Tuple, cast
+
+import torch
+from torch._inductor.codecache import FxGraphCachePickler
+from torch.fx.experimental.proxy_tensor import maybe_disable_fake_tensor_mode
+
+_LOGGER: logging.Logger = logging.getLogger(__name__)
+
+
+class BaseEngineCache(ABC):
+
+    @abstractmethod
+    def __init__(
+        self,
+        *args: Any,
+        **kwargs: Any,
+    ) -> None:
+        pass
+
+    @staticmethod
+    def get_hash(gm: torch.fx.GraphModule) -> str:
+        """Get the hash value of the GraphModule
+
+        Args:
+            gm (torch.fx.GraphModule): GraphModule to hash
+
+        Returns:
+            str: hash value of the GraphModule
+        """
+        # parameters are set to 0
+        with maybe_disable_fake_tensor_mode():
+            new_gm = copy.deepcopy(gm)
+            for name, param in new_gm.named_parameters():
+                param.data.zero_()
+
+            hash_val = cast(str, FxGraphCachePickler.get_hash(new_gm))
+
+        return hash_val
+
+    @abstractmethod
+    def save(
+        self,
+        hash: str,
+        serialized_engine: bytes,
```

Comment:
Do you think we should give people a single blob to save with all the info or have them pack it themselves? This blob could be a pkl or something that is easy to store in a bin file or db

Reply:
We can define the packing as a static method of the base class where `(serialized_engine, input_names, output_names, weight_map) -> bytes` then this function can be `def save(self, hash, blob) -> bool`